docs: add scaling guide for crawlers (#169)
apify/apify-ts: docs/guides/scaling_crawlers.mdx

import AutoscaledPoolOptionsSource from '!!raw-loader!./scaling_crawlers_autoscaledPoolOptions.ts';

As you build your crawler, you might want to control how many requests you do to the website at a time. We provide several options to fine tune how many parallel requests should be made at a time, how many requests should be done per minute, and how should scaling up and down happen based on the system resources available.

:::tip

All of these options are available on all crawlers we provide, but for this guide we'll be using the <ApiLink to="cheerio-crawler/class/CheerioCrawler">`CheerioCrawler`</ApiLink>. You can see all options that are available <ApiLink to="cheerio-crawler/interface/CheerioCrawlerOptions">`here`</ApiLink>
:::

## `minConcurrency` and `maxConcurrency`

These control how many parallel requests can be run at a time. By default, crawlers will start with one parallel request at a time and scale up over time to a maximum of 1000 requests at a time.

:::caution Don't set `minConcurrency` too high!

Setting this option too high compared to the available system resources will make your crawler run extremely slow or might even crash.

> We recommend you leave it at the default value we set and let the crawlers scale up and down automatically based on available resources instead.

Author: Rewording this sentence to made a distinction between "we" the guide reader and "we" crawlee was fun... could you check it now and tell me if it's better? 😅

> We recommend you leave it at the default value we set and let the crawlers scale up and down automatically based on available resources instead.

Reviewer: I guess should be good :) And tbh, I still have several places where I also need to do the same thing :)

:::tip

All of these options are available on all crawlers Crawlee provides, but for this guide we'll be using the <ApiLink to="cheerio-crawler/class/CheerioCrawler">`CheerioCrawler`</ApiLink>. We can see all options that are available <ApiLink to="cheerio-crawler/interface/CheerioCrawlerOptions">`here`</ApiLink>
:::

## `minConcurrency` and `maxConcurrency`

These control how many parallel requests can be run at any time. By default, crawlers will start with one parallel request at a time and scale up over time to a maximum of 1000 requests at a time.

:::caution Don't set `minConcurrency` too high!

Setting this option too high compared to the available system resources will make our crawler run extremely slow or might even crash.

It's recommended to leave it at the default value that is provided and letting the crawler scale up and down automatically based on available resources instead.

:::

<CodeBlock language="js">
	{MinMaxConcurrencySource}
</CodeBlock>

> ## `maxRequestsPerMinute`

Reviewer 2: I would place this above the min/maxConcurrency as it's a much better rate limiting option than those two.

I'm also thinking that maybe we should set some reasonable default, like 200 or something that does not overload a website 😄

Reviewer 3: Makes sense. On the other hand, if we have `maxConcurrency` set to 1000 as the paragraph above states, maybe we should adjust that as well?

Author: We could change the default for maxConcurrency to 200 from 1000, I think that's a smart idea. I'll also move this above them as I agree with you

This controls how many total requests can be made per minute. It counts the amount of requests done every second, to ensure there is not a burst of requests at the `maxConcurrency` limit followed by a long period of waiting. By default, it is set to `Infinity` which means the crawler will keep going up to the `maxConcurrency`. We would set this if we wanted our crawler to work at full throughput, but also not keep hitting the website we're crawling with non-stop requests.

<CodeBlock language="js">
	{MaxRequestsPerMinuteSource}
</CodeBlock>

## Advanced configurations

While the options above should be enough for most users, if we wanted to get super deep into the configuration of the autoscaling pool (the internal utility in Crawlee that helps us allow crawlers to scale up and down), we can do so through the <ApiLink to="cheerio-crawler/interface/CheerioCrawlerOptions#autoscaledPoolOptions">`autoscaledPoolOptions`</ApiLink> object available on crawler options.

:::danger Complex options up ahead!

> This section is super advanced and, unless we test the changes extensively and know what we're doing, it's better to leave these options to the default configurations provided, as they are most likely going to work for us without much fuss.

Reviewer 3: while generally i want to see the "we" in docs, here (or in general in hints like this) i'd say its better to keep using the "you". as the docs should be about telling a story, but hints are like "hey you, heads up!"

```suggestion
This section is super advanced and, unless you test the changes extensively and know what you're doing, it's better to leave these options to their defaults, as they are most likely going to work fine without much fuss.
```

:::

With that warning aside, if we're feeling adventurous, this is how we would pass these options when using a crawler:

<CodeBlock language="js">
	{AutoscaledPoolOptionsSource}
</CodeBlock>

### `desiredConcurrency`

This option specifies the amount of requests that should be running in parallel at the start of the crawler, assuming there are so many available. It defaults to the same value as `minConcurrency`.

### `desiredConcurrencyRatio`

The minimum ratio of concurrency to reach before more scaling up is allowed (a number between `0` and `1`). By default, it is set to `0.95`.

We can think of this as the point where the autoscaling pool can attempt to scale up (or down), monitor if there's any changes, and correct them if necessary.

### `scaleUpStepRatio` and `scaleDownStepRatio`

These values define the fractional amount of desired concurrency to be added or subtracted as the autoscaling pool scales up or down. Both of these values default to `0.05`.

:::caution

> If we're planning on modifying these, make sure the value we set is not `0`, and that it is positive!

Author: We don't actually have strict validation for this... 😓  Would probably help to fix that too
